app/src/services/kleros.ts

```diff
@@ -0,0 +1,133 @@
+]
+
+class KlerosService {
+  provider: Web3Provider
+  badgeContract: Contract
+  tokensViewContract: Contract
+
+  constructor(
+    badgeContractAddress: string,
+    tokensViewContractAddress: string,
+    provider: Web3Provider,
+    signerAddress: Maybe<string>,
+  ) {
+    this.provider = provider
+    if (signerAddress) {
+      const signer = provider.getSigner()
+      this.badgeContract = new ethers.Contract(
+        badgeContractAddress,
+        klerosBadgeAbi,
+        provider,
+      ).connect(signer)
+      this.tokensViewContract = new ethers.Contract(
+        tokensViewContractAddress,
+        klerosTokensViewAbi,
+        provider,
+      ).connect(signer)
+    } else {
+      this.badgeContract = new ethers.Contract(badgeContractAddress, klerosBadgeAbi, provider)
+      this.tokensViewContract = new ethers.Contract(
+        tokensViewContractAddress,
+        klerosTokensViewAbi,
+        provider,
+      )
+    }
+  }
+
+  /**
+   * Return the tokens from the KLEROS TCR already registered
+   * @returns {Promise<any[]>}
+   */
+  queryTokens = async () => {
+    const network = await this.provider.getNetwork()
+    const networkId = network.chainId
+
+    if (networkId !== networkIds.MAINNET) {
+      // Use mocked information from networks file
+      return getTokensByNetwork(networkId)
+    } else {
+      const filter = [
+        false, // Include absent tokens in result
+        true, // Include registered tokens in result.
+        false, // Include tokens with registration requests that are not disputed in result.
+        true, // Include tokens with clearing requests that are not disputed in result.
+        false, // Include disputed tokens with registration requests in result.
+        true, // Include disputed tokens with clearing requests in result.
+        false, // Include tokens submitted by the caller.
+        false, // Include tokens challenged by the caller.
+      ]
+
+      let hasMore = true
+      let addressesWithBadge: string[] = []
+
+      // Fetch addresses of tokens that have the badge.
+      // Since the contract returns fixed sized arrays, we must filter out unused items.
+      while (hasMore) {
+        const result = await this.badgeContract.queryAddresses(
+          ethers.constants.AddressZero,
+          1000,
+          filter,
+          true,
+        )
+        addressesWithBadge = addressesWithBadge.concat(
+          result[0].filter((address: string) => address !== ethers.constants.AddressZero),
+        )
+        hasMore = result.hasMore
+      }
+
+      // Fetch their submission IDs on the TCR.
+      const network = await this.provider.getNetwork()
+      const networkId = network.chainId
+      const klerosTCRAddress = getContractAddress(networkId, 'klerosTCR')
```

Comment:
I would add this as a constructor parameter and then use it as an instance variable.

Reply:
Done!